### PR TITLE
remount volumes on restart

### LINKED ISSUE #4264: empty volumes when using overlay on Prod
We have been seeing builds error out more frequently with some common error messages lately:

`task config 'ci/tasks/yarn-build.yml' not found` -> https://ci.concourse-ci.org/teams/main/pipelines/prs/jobs/watsjs/builds/1574

`file not found` -> https://ci.concourse-ci.org/teams/main/pipelines/prs/jobs/testflight/builds/1569

Cert errors on checks - 
```resource script '/opt/resource/check []' failed: exit status 1

stderr:
[31mERRO[0m[0000] failed to get remote image: Get https://index.docker.io/v2/: x509: certificate signed by unknown authority
```

@kcmannem and I theorize that this has to do with something that has recently changed with volumes -- perhaps `overlay`? streaming? A combination of the two?

We're not sure these issues are all related but this issue exists to document the work to investigate these errors.

We've been seeing these since around the time we pushed Concourse v5.4.0.

### PR DESCRIPTION
**Pull request by kcmannem**
Fixes https://github.com/concourse/concourse/issues/4264

Signed-off-by: Krishna Mannem <kmannem@pivotal.io>

### CODE REVIEW

File: volume/driver/overlay_linux.go [hunk 1]
```diff
@@ -30,7 +42,9 @@ func (driver *OverlayDriver) CreateVolume(path string) error {
 
 func (driver *OverlayDriver) DestroyVolume(path string) error {
 	err := syscall.Unmount(path, 0)
-	if err != nil {
+	// when a path is already unmounted, and unmount is called
+	// on it, syscall.EINVAL is returned as an error
```
Comment (ddadlani) [inline volume/driver/overlay_linux.go:51]: How are `syscall.EINVAL` and `os.ErrInvalid` related?

File: volume/driver/overlay_linux.go [hunk 2]
```diff
@@ -78,21 +92,99 @@ func (driver *OverlayDriver) CreateCopyOnWriteLayer(path string, parent string)
 		workDir,
 	)
 
+	fmt.Println(path, opts)
```
Comment (ddadlani) [inline volume/driver/overlay_linux.go]: not sure we need this print line?

File: volume/driver/overlay_linux.go [hunk 3]
```diff
@@ -2,16 +2,28 @@ package driver
...
+	VolumesDir string
 	OverlaysDir string
 }
 
+func NewOverlayDriver(volumesDir, overlaysDir string) *OverlayDriver {
+	driver := &OverlayDriver{
+		VolumesDir:  volumesDir,
+		OverlaysDir: overlaysDir,
+	}
+	driver.RecoverMountTable(filepath.Join(volumesDir, "live"))
```
Comment (cirocosta) [inline volume/driver/overlay_linux.go]: Are we ignoring the error from `RecoverMountTable` on purpose here? 🤔

File: volume/driver/overlay_linux.go (hunk 2 above)
Comment (cirocosta) [inline volume/driver/overlay_linux.go]: I guess this is debugging a leftover 🤔

File: volume/driver/overlay_linux.go [hunk 4]
```diff
@@ -78,21 +92,99 @@ func (driver *OverlayDriver) CreateCopyOnWriteLayer(path string, parent string)
...
+	}
+
+	for _, volumeFileInfo := range(liveVolumes) {
+
+		liveVolumePath := filepath.Join(liveVolumesDir, volumeFileInfo.Name())
+		liveVolumeDataPath := filepath.Join(liveVolumePath, "volume")
+		parentSymlink := filepath.Join(liveVolumePath, "parent")
+
+		// a parent symlink indicates a COW
+		if _, err := os.Stat(parentSymlink); !os.IsNotExist(err) {
```
Comment (cirocosta) [inline volume/driver/overlay_linux.go]: I'm a bit confused with this one 🤔 

Does this mean that "for any `err` that is not `ENOENT`, go ahead with the following"? I.e., "go ahead for EACCES, ENAMETOOLONG, ENOMEM, etc etc"

File: volume/driver/overlay_linux.go (hunk 3 above)
Comment (kcmannem) [reply volume/driver/overlay_linux.go]: nope, I missed it. thanks.

File: volume/driver/overlay_linux.go (hunk 4 above)
Comment (kcmannem) [reply volume/driver/overlay_linux.go]: Yeah it does, I placed it here as a safety check to see that the parent symlink exists. The symlink is always named "parent" and is created by the baggage claim process. So I think we're cool about permission denied or name too long errors. If you run out of memory you probably have bigger problems than this. I think it's reasonable to assume that IsExists is the only error to care for?

File: volume/driver/overlay_linux.go (hunk 2 above)
Comment (kcmannem) [reply volume/driver/overlay_linux.go]: nah, this is a crucial step for booting up baggageclaim.

File: volume/driver/overlay_linux.go (hunk 4 above)
Comment (kcmannem) [reply volume/driver/overlay_linux.go]: fixed lemme know what you think

File: volume/driver/overlay_linux.go (hunk 2 above)
Comment (ddadlani) [reply volume/driver/overlay_linux.go]: you took care of this, nvm

**Comment by ddadlani** [comment]
@xtreme-sameer-vohra and I tried out this PR and it seems to work. This was our process:

**Acceptance** 
1) Run Concourse with one worker using `docker-compose`
2) Set a pipeline and watch it run successfully
3) Run `docker restart concourse_worker_1` and re-run the pipeline

Without this PR, we saw that `booklit/unit` would error out with `task config not found`, implying that the booklit volume was empty. We also double checked on the worker, that the volume corresponding to booklit was empty but the overlay directory had all the bits in it.

With the PR, we saw that the task ran successfully, and that the volume corresponding to booklit was correctly populated, and that `cat /proc/mounts` showed the corresponding overlay mount on the worker.

**Asides**
1) In both the passing and failing cases, we saw `check` containers failing with `unknown handle` because the ATC was not yet aware that the containers were no longer present on the worker. This resolves itself once the `missing_since` column in the containers table triggers those containers to be GCed, but this is not a clear error.
2) To test this case, we needed to disable worker retiring. This is because, when running `docker restart`, the worker transitions to retiring and then back to running. This caused the volumes to get GCed from the DB, but the worker restarted too quickly for the actual volumes to be cleaned up from the worker. This meant that we did not use the existing volumes on the worker, and also that every restart caused volumes to pile up. However this is specific to using `docker restart` as workers do not transition from retiring to running.

File: volume/driver/overlay_linux.go (hunk 1 above)
Comment (kcmannem) [reply volume/driver/overlay_linux.go:51]: Don't know why both exist in the stdlib. This is some go wierdness.
```
fmt.Println(syscall.EINVAL.Error()) // Invalid argument
fmt.Println(os.ErrInvalid) // invalid argument
```

The error you get is lower case so i used os.ErrInvalid
```
"data":{"error":"invalid argument","handle":"some-cow-handle","session":"3.1.4.1"}
```

Comment (ddadlani) [reply volume/driver/overlay_linux.go:51]: hmm that is weird. I did also see that in the `os` package we return ErrInvalid so I'm ok with sticking with this for now.

**Pull request merged**